lib/Dialect/QUIR/Transforms/BreakReset.cpp

```diff
@@ -131,13 +143,149 @@ void BreakResetPass::runOnOperation() {
   // Disable to improve performance
   config.enableRegionSimplification = false;
 
-  patterns.add<BreakResetsPattern>(&getContext(), numIterations, delayCycles);
+  patterns.add<BreakResetsPattern>(&getContext(), numIterations, delayCycles,
+                                   *this);
 
   if (mlir::failed(applyPatternsAndFoldGreedily(getOperation(),
                                                 std::move(patterns), config)))
     signalPassFailure();
+
+  // put measures and call gates into circuits -- when
+  // putCallGatesAndMeasuresIntoCircuit option is true
+  mlir::ModuleOp moduleOp = getOperation();
+  uint circNum = 0;
+  while (!measureList.empty()) {
+    MeasureOp measOp = dyn_cast<MeasureOp>(measureList.front());
+    putMeasureInCircuit(moduleOp, measOp, circNum);
+    circNum++;
+    measureList.pop_front();
+  }
+  while (!callGateList.empty()) {
+    CallGateOp callGateOp = dyn_cast<CallGateOp>(callGateList.front());
+    putCallGateInCircuit(moduleOp, callGateOp, circNum);
+    circNum++;
+    callGateList.pop_front();
+  }
 } // BreakResetPass::runOnOperation
 
+void BreakResetPass::putCallGateInCircuit(ModuleOp moduleOp,
+                                          mlir::quir::CallGateOp callGateOp,
+                                          uint circNum) {
+  mlir::func::FuncOp mainFunc =
+      dyn_cast<mlir::func::FuncOp>(quir::getMainFunction(moduleOp));
+  assert(mainFunc && "could not find the main func");
+  mlir::OpBuilder builder(mainFunc);
+  mlir::Location location = mainFunc.getLoc();
+
+  auto circOp = builder.create<CircuitOp>(
+      mainFunc.getLoc(), "reset_x_" + std::to_string(circNum++),
```

Comment:
What happens if a circuit with this name already exists?

Reply:
Yes, given that the compiler takes milr as input I don't think we can assume that a circuit that we are creating is unique I think we should test for the existence of a circuit name before creating the new one and adjust the new one if required.